tls/s2n_handshake_io.c

```diff
@@ -1079,10 +1086,16 @@ int s2n_negotiate(struct s2n_connection *conn, s2n_blocked_status * blocked)
             }
         } else {
             *blocked = S2N_BLOCKED_ON_READ;
+
             int r = handshake_read_io(conn);
 
             if (r < 0) {
-                s2n_try_delete_session_cache(conn);
+                /* One blocking condition is waiting on the session resumption cache. */
+                /* So we don't want to delete anything if we are blocked. */
+                if (s2n_errno != S2N_ERR_BLOCKED && s2n_allowed_to_cache_connection(conn) && conn->session_id_len) {
```

Comment:
address of array 'conn->session_id' will always evaluate to 'true'

Reply:
sorry, this is referring to line 895

Reply:
Good catch, this was a copy/paste error. This should be conn->session_id_len. Updated.

Reply:
Updated.